### PR TITLE
[mlir][transform] Add transform.get_operand op

### PR DESCRIPTION
**Pull request by qedawkins**
Similar to `transform.get_result`, except it returns a handle to the operand indicated by `operand_number`, or all operands if no index is given.

Additionally updates `get_result` to make the `result_number` optional. This makes the use case of wanting to get all of the results of an operation easier by no longer requiring the user to reconstruct the list of results one-by-one.

### CODE REVIEW

File: mlir/lib/Dialect/Transform/IR/TransformOps.cpp [hunk 1]
```diff
@@ -1464,6 +1464,35 @@ transform::GetProducerOfOperand::apply(transform::TransformRewriter &rewriter,
...
+DiagnosedSilenceableFailure
+transform::GetOperandOp::apply(transform::TransformRewriter &rewriter,
+                               transform::TransformResults &results,
+                               transform::TransformState &state) {
+  std::optional<int64_t> maybeOperandNumber = getOperandNumber();
+  SmallVector<Value> operands;
+  for (Operation *target : state.getPayloadOps(getTarget())) {
+    if (!maybeOperandNumber) {
+      for (Value operand : target->getOperands())
+        operands.push_back(operand);
```
Comment (ftynse) [inline mlir/lib/Dialect/Transform/IR/TransformOps.cpp:1480]: Nit: `llvm::append_range` should work here, and will avoid the potential multiple-reallocation issue due to appending to vector in a loop.

**Review by ftynse — APPROVED** [review]
Looks good in principle. 

An overall design comment: we have a similar mechanism in Linalg matcher operations, e.g. https://mlir.llvm.org/docs/Dialects/Transform/#transformmatchstructuredinit-transformmatchstructuredinitop. It has more advanced addressing modes, e.g., associating a list of operands with the result. Not necessarily arguing to adopt all of that here immediately or at all, but I'd recommend using an explicit token/attribute to indicate that `all` operands are requested instead of omitting the attribute.

File: mlir/lib/Dialect/Transform/IR/TransformOps.cpp [hunk 2]
```diff
@@ -1464,6 +1464,35 @@ transform::GetProducerOfOperand::apply(transform::TransformRewriter &rewriter,
...
+      continue;
+    }
+    int64_t operandNumber = *maybeOperandNumber;
+    if (operandNumber >= target->getNumOperands()) {
+      DiagnosedSilenceableFailure diag =
+          emitSilenceableError() << "targeted op does not have enough operands";
```
[the remark this answers is not in the document]
Comment (qedawkins) [reply mlir/lib/Dialect/Transform/IR/TransformOps.cpp]: This is resolved by switching to the `StructuredTransformDims` approach.

**Comment by qedawkins** [comment]
> Looks good in principle.
> 
> An overall design comment: we have a similar mechanism in Linalg matcher operations, e.g. https://mlir.llvm.org/docs/Dialects/Transform/#transformmatchstructuredinit-transformmatchstructuredinitop. It has more advanced addressing modes, e.g., associating a list of operands with the result. Not necessarily arguing to adopt all of that here immediately or at all, but I'd recommend using an explicit token/attribute to indicate that `all` operands are requested instead of omitting the attribute.

Good suggestion, I like the linalg matcher approach more so I just imported it. I'll leave this open for a day or so to give a chance to review if you want to check that I got the code shuffling right.

**Review by ftynse — APPROVED** [review]

**Pull request merged**